File: src/main/com/jatti/achievements/prize/Prize.java
package com.jatti.achievements.prize;

import org.bukkit.inventory.ItemStack;

import java.util.ArrayList;
import java.util.List;

public class Prize {

    private int value;
    private double xp;
    private int gold;
    private List<ItemStack> items;

    public Prize(int value) {

        this.value = value;
        PrizeUtils.addPrize(this);

    }

    public static Prize get(int value) {

        for (Prize p : PrizeUtils.getPrizes()) {

            if (p.getValue() == value) return p;

        }

        return new Prize(value);

    }

    public List<ItemStack> getItems() {
        if (items == null) items = new ArrayList<ItemStack>();
        return items;
    }

    public void setItems(List<ItemStack> items) {
        this.items = items;
    }

    public int getGold() {
        return gold;
    }

    public void setGold(int gold) {
        this.gold = gold;
    }

    public double getXp() {
        return xp;
    }

    public void setXp(double xp) {
        this.xp = xp;
    }

    public int getValue() {
        return value;
    }

    public void setValue(int value) {
        this.value = value;
    }
}

File: src/main/com/jatti/achievements/prize/Prize.kt
package com.jatti.achievements.prize

import com.jatti.user.User
import org.bukkit.Sound
import org.bukkit.inventory.ItemStack

/**
 * Prize e.g for completing mission or getting achievement
 * @author Jatti
 * @version 1.0
 */
open class Prize {

    var items: List<ItemStack>? = ArrayList()
    var gold: Int = 0
    var exp: Double = 0.0
    var user: User? = null
    var sound: Sound? = null

    companion object {
        /**
         * Creates new builder to make Prize
         */
        @JvmStatic
        fun builder(): PrizeBuilder {
            return PrizeBuilder()
        }
    }

}

/**
 * Builder for Prize
 * @author Jatti
 * @version 1.0
 */
class PrizeBuilder {

    private val p = Prize()

    /**
     * Addes items to Prize
     * @param items list of item
     */
    fun withItems(items: List<ItemStack>): PrizeBuilder {
        p.items = items
        return this
    }

    /**
     * Addes gold to Prize
     * @param gold amount of gold
     */
    fun withGold(gold: Int): PrizeBuilder {
        p.gold = gold
        return this
    }

    /**
     * Addes exp to Prize
     * @param exp amount of exp
     */
    fun withExp(exp: Double): PrizeBuilder {
        p.exp = exp
        return this
    }

    fun withSound(sound: Sound): PrizeBuilder {
        p.sound = sound
        return this
    }
    /**
     * definies user which will get prize (MUST HAVE)
     * @param user user which will get prize
     */
    fun forUser(user: User): PrizeBuilder {
        p.user = user
        return this
    }

    /**
     * Creates Prize and gives it to user
     */
    fun build(): Prize {

        p.user!!.getPlayer().playSound(p.user!!.getPlayer().location, p.sound!!, 0F, 1F)

        if (p.items != null) {
            for (i in p.items!!) {
                p.user!!.getPlayer().inventory.addItem(i)
            }
        }

        p.user!!.addGold(p.gold)
        p.user!!.addExp(p.exp)

        return p

    }

}